File: Python/Gaus/gaus.py

with open ('exit.txt', 'w') as exit:
        exit.write("")


def gaus(n, matr):
    with open ('exit.txt', 'a') as exit:
        x = [0 for i in range(n) ]
        mat = matr
        for i in range(n):
            if mat[i][i] == 0:
                exit.write("There is no solution")
                break
            for j in range(i+1, n): 
                tem = mat[j][i]/mat[i][i]

                for k in range(n+1):
                    mat[j][k] = mat[j][k] - tem * mat[i][k]

        x[n-1] = mat[n-1][n]/mat[n-1][n-1]

        for i in range(n-2,-1,-1):
            x[i] = mat[i][n]

            for j in range(i+1,n):
                x[i] = x[i] - mat[i][j]*x[j]

            x[i] = x[i]/mat[i][i]

        for i in range(n):
            exit.write(f"X{i+1} = "+str(round(x[i], 4)))
            exit.write('\t')
        exit.write('\n')
        
        

def call():
        matrix = []
        with open('input.txt', 'r') as inpt:
                for line in inpt:               
                        if line != "\n" :
                                t = len(line.split())-1
                                curr = []
                                matrix.append(list(map(int, line.split())))
                        else:
                                gaus(t, matrix)
                                matrix = []    
call()

File: cpp/gaus/gaus.cpp
#include "../matrix.hpp"
#include "gaus.h"
#include <string>
#include <regex>

const std::string inp = "../gaus/input.txt";
const std::string gold = "../gaus/golden.txt";

bool validate(std::string file){
        std::ifstream input(file);
        std::string line;
        std::regex p("([0-9]{1}) ([0-9]{1})");
        std::smatch m;
        std::regex l("(\\w+)");
        std::smatch c;
        std::string temp;
        int iter;
        if(!input.good()) {
                input.close();
                return false;
        }
        if (input.peek() == std::ifstream::traits_type::eof()){
                input.close();
                return false;
        }
        for (int lineNum = 1; getline(input, line);lineNum++){
            if (regex_search(line, m, p) && line.length()<6){
                std::cout << std::endl;
                if(std::stoi(m.str(1))+1 == std::stoi(m.str(2))){
                        int row = std::stoi(m.str(1));
                        int column = std::stoi(m.str(2));
                        iter = column*row;
                        std::cout<<"\n"<<row<<"\t"<<column<<"\n";
                        while (iter--) {
                                input>>temp;
                                std::cout<<temp<<" ";
                                if (!std::stoi(temp)) {
                                        input.close();
                                        return false;
                                }
                                std::cout<<"\n";
                        }
                } else {
                        input.close();
                        return false;
                }
            }
            else if (regex_search(line,c,l)) {
                std::cerr << "There's an invalid entry not Number!" << std::endl;
                //input.close();
                //return false;
            }
        }
        input.close();
        return true;
}

void test()
{
        std::ifstream input(inp);
        std::ifstream golden(gold);
        std::ofstream exit("exit.txt");
        std::ofstream result("result.txt");
        int a,b;
        int count = 0;
        while (!input.eof()) {
                input>>a>>b;
                if (input.eof()) {
                        break;
                }
                Matrix matrix(a,b);
                input>>matrix;
                float* x = gaus(matrix);
                float* y = new float[a];
                count++;
                exit<<std::endl;
                for (int i = 0;i<a;i++) {
                        exit<<"X"<<i+1<<" = "<<x[i]<<"\t";
                        golden>>y[i];
                        if (std::abs(x[i]-y[i])>0.001) {
                                std::cout<<x[i]<<"\t"<<y[i]<<std::endl;
                                result<<"Test "<<count<<" Failed! your result: "<<x[i]<<"  expected result: "<<y[1]<<"\n";
                                break;
                        } else if (i == (a-1) && x[i] == y[i]) {
                                result<<"Test "<<count<<" Passed Succsesfully\n";
                        }
                }

        }
        input.close();
        golden.close();
        exit.close();
        result.close();
}

int main()
{
	//test();
        std::cout<<std::endl<<validate(inp)<<std::endl;
	return 0;
}
